retrofit/src/main/java/retrofit2/HttpServiceMethod.java

```diff
@@ -31,13 +36,32 @@
    */
   static <ResponseT, ReturnT> HttpServiceMethod<ResponseT, ReturnT> parseAnnotations(
       Retrofit retrofit, Method method, RequestFactory requestFactory) {
-    CallAdapter<ResponseT, ReturnT> callAdapter = createCallAdapter(retrofit, method);
-    Type responseType = callAdapter.responseType();
-    if (responseType == Response.class || responseType == okhttp3.Response.class) {
+    CallAdapter<ResponseT, ReturnT> callAdapter = null;
+    boolean continuationWantsResponse = false;
+    Type responseType;
+    if (requestFactory.isKotlinSuspendFunction) {
+      Type[] parameterTypes = method.getGenericParameterTypes();
+      Type continuationType = parameterTypes[parameterTypes.length - 1];
+      responseType = Utils.getParameterLowerBound(0, (ParameterizedType) continuationType);
+      if (getRawType(responseType) == Response.class && responseType instanceof ParameterizedType) {
+        // Unwrap the actual body type from Response<T>.
+        responseType = Utils.getParameterUpperBound(0, (ParameterizedType) responseType);
+        continuationWantsResponse = true;
+      }
+    } else {
+      callAdapter = createCallAdapter(retrofit, method);
+      responseType = callAdapter.responseType();
+    }
```

Comment:
Unfortunately this part avoid the usage of custom `CallAdapter` together with `suspended` functions.

I played a bit around and it looks like the feature to skip the `Response` class together with `suspended` functions makes it necessary to skip `createCallAdapter()` calls.

When I remove the skip `Response` class feature and change some parts then I can use custom `CallAdapter` again.

```
    if (requestFactory.isKotlinSuspendFunction) {
      Type[] parameterTypes = method.getGenericParameterTypes();
      Type continuationType = parameterTypes[parameterTypes.length - 1];
      responseType = Utils.getParameterLowerBound(0, (ParameterizedType) continuationType);
      if (getRawType(responseType) == Response.class && responseType instanceof ParameterizedType) {
        // Unwrap the actual body type from Response<T>.
        responseType = Utils.getParameterUpperBound(0, (ParameterizedType) responseType);
        continuationWantsResponse = true;
      } else {
        callAdapter = createCallAdapter(retrofit, method, responseType);
        responseType = callAdapter.responseType();
      }
    } else {
        callAdapter = createCallAdapter(retrofit, method, method.getGenericReturnType());
        responseType = callAdapter.responseType();  
    }
```
```
  private static <ResponseT, ReturnT> CallAdapter<ResponseT, ReturnT> createCallAdapter(
      Retrofit retrofit, Method method, Type returnType) {
    Annotation[] annotations = method.getAnnotations();
    try {
      //noinspection unchecked
      return (CallAdapter<ResponseT, ReturnT>) retrofit.callAdapter(returnType, annotations);
    } catch (RuntimeException e) { // Wide exception range because factories are user code.
      throw methodError(method, e, "Unable to create call adapter for %s", returnType);
    }
  }
```
I have just remove the failing tests (mostly related to the skipping `Response` class feature) for creating the package.
Perhaps the skip `Response` class feature could be delivered as a plugin?

If someone want to try it
* `git clone https://github.com/nenick/retrofit.git`
* `git checkout jakew/suspend/2018-09-11`
* `mvn package`
* replace your retrofit dependency with `retrofit/retrofit/target/retrofit-2.4.1-SNAPSHOT`

Reply:
I created an example how it could look to support custom `CallAdapter` for the suspend feature. Main goal is to hide Retrofit details from clients and use some custom error handling.
https://github.com/nenick/retrofit/tree/suspend-calladapter

Mainly I refactored all call adapting logic into `SuspendCallAdapter` classes and provide a `DefaultSuspendCallAdapterFactory` to keep @JakeWharton default implementations. So we can decide whether to use the `DefaultSuspendCallAdapterFactory` or provide my own custom handling. 

*all existing tests are green*

**Please tell if your want this feature added. Then I try to create a clean PullRequest with your help.**

Usage:

```
Retrofit.Builder()
        .baseUrl(/* server url */)
        .addSuspendCallAdapterFactory(DefaultSuspendCallAdapterFactory())
        .build()
```

*Hint: For implementing your own Factory&Adapter take existing ones as a template.*

Reply:
I was talking about that above, but Jake said it is out of scope for this MR https://github.com/square/retrofit/pull/2886#issuecomment-438936312

Reply:
I didn't wanted to wait any longer or creating some ugly workarounds. And it is not clear why there is no progress on this PR. If help is wanted then tell it.

If someone want then he can use my custom release artifact https://github.com/nenick/retrofit/releases/tag/2.5.1-coroutines
until this PR is merged and we could clean up this hacky code. I use it by my self and it works pretty well.

Reply:
Finally, thanks!

Reply:
@nenick Any chance you can submit PR for this functionality?